packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -1,242 +1,274 @@
-      getNetworkState: () => NetworkState;
-      getPreferencesState: () => PreferencesState;
-      getNetworkClientById: NetworkController['getNetworkClientById'];
-    },
-    config?: Partial<TokenDetectionConfig>,
-    state?: Partial<BaseState>,
-  ) {
-    const {
-      providerConfig: { chainId: defaultChainId },
-    } = getNetworkState();
+  constructor({
+    chainId,
+    selectedAddress = '',
+    interval = DEFAULT_INTERVAL,
+    onPreferencesStateChange,
+    getBalancesInSingleCall,
+    addDetectedTokens,
+    getTokensState,
+    getPreferencesState,
+    messenger,
+  }: {
+    chainId: Hex;
+    selectedAddress?: string;
+    interval?: number;
+    onPreferencesStateChange: (
+      listener: (preferencesState: PreferencesState) => void,
+    ) => void;
+    getBalancesInSingleCall: AssetsContractController['getBalancesInSingleCall'];
+    addDetectedTokens: TokensController['addDetectedTokens'];
+    getTokensState: () => TokensState;
+    getPreferencesState: () => PreferencesState;
+    messenger: TokenDetectionControllerMessenger;
+  }) {
     const { useTokenDetection: defaultUseTokenDetection } =
       getPreferencesState();
 
-    super(config, state);
-    this.defaultConfig = {
-      interval: DEFAULT_INTERVAL,
-      selectedAddress: '',
-      disabled: true,
-      chainId: defaultChainId,
-      isDetectionEnabledFromPreferences: defaultUseTokenDetection,
-      isDetectionEnabledForNetwork:
-        isTokenDetectionSupportedForNetwork(defaultChainId),
-      ...config,
-    };
-
-    this.initialize();
-    this.setIntervalLength(this.config.interval);
-    this.getTokensState = getTokensState;
-    this.getTokenListState = getTokenListState;
-    this.addDetectedTokens = addDetectedTokens;
-    this.getBalancesInSingleCall = getBalancesInSingleCall;
-    this.getNetworkClientById = getNetworkClientById;
-
-    onTokenListStateChange(({ tokenList }) => {
-      const hasTokens = Object.keys(tokenList).length;
-
-      if (hasTokens) {
-        this.detectTokens();
-      }
+    super({
+      name: controllerName,
+      messenger,
+      state: {},
+      metadata: {},
     });
 
-    onPreferencesStateChange(({ selectedAddress, useTokenDetection }) => {
-      const {
-        selectedAddress: currentSelectedAddress,
-        isDetectionEnabledFromPreferences,
-      } = this.config;
-      const isSelectedAddressChanged =
-        selectedAddress !== currentSelectedAddress;
-      const isDetectionChangedFromPreferences =
-        isDetectionEnabledFromPreferences !== useTokenDetection;
-
-      this.configure({
-        isDetectionEnabledFromPreferences: useTokenDetection,
-        selectedAddress,
-      });
+    this.#disabled = false;
+    this.setIntervalLength(interval);
+    this.#chainId = chainId;
+    this.#selectedAddress = selectedAddress;
+    this.#networkClientId = this.messagingSystem.call(
+      'NetworkController:findNetworkClientIdByChainId',
+      chainId,
+    );
```

Comment:
Applied here: https://github.com/MetaMask/core/compare/173168f172ec1b1b56139c6a7ecb559c49ad7c36...1b109d2bafd58658200c2811c9837a019ae9fedb

This required a number of test fixes, and updating `TokensController` to rely on its controller-messenger for `onNetworkDidChange`, `TokenListStateChange` events. This was necessary so that those events being triggered in `TokenDetectionController` tests would update `TokensController` state.

I've fixed all but two failing tests -- one in `TokenDetectionController` and one in `TokensController`: https://github.com/MetaMask/core/actions/runs/7262601520/job/19786134002